### PR TITLE
Add Spark 4.1.2 shim support

### LINKED ISSUE #14832: [FEA] Add support for Spark 4.1.2
**Is your feature request related to a problem? Please describe.**
N/A

**Describe the solution you'd like**
Add a shim layer to support Spark 4.1.2

**Describe alternatives you've considered**
N/A

**Additional context**
https://lists.apache.org/thread/locxv3h4tkbsvcf8t600kyvx16xw6xl3

### PR DESCRIPTION
**Pull request by firestarman**
Fixes #14832.
Contributes to https://github.com/NVIDIA/cudf-spark/issues/15073

### Description

- Add Spark 4.1.2 build profiles and generated shim metadata so the Scala 2.13 build can produce a Spark 4.1.2 RAPIDS jar.
- Add Spark 4.1.2 shim sources and extract shared Spark 4.1.x rules into `Spark411PlusShims` to avoid duplicating common 4.1.x expression and execution registrations.
- Add Spark 4.1.2-specific ACOSH/ASINH compatibility expressions so GPU results match Spark CPU default behavior for large-input and invalid-domain cases.
- Update AST integration test expectations for ACOSH on Spark 4.1.2+ because the compatibility expression requires branch logic that is not AST-compatible.
- Validation run locally with `SPARK_HOME=/bigdata/work/tools/spark-4.1.2-bin-hadoop3 mvn -s /home/liangcail/.m2/settings_art.xml -Dcuda.version=cuda13 -f scala2.13/pom.xml -Dbuildver=412 clean verify`.
- Local NDS performance results show no overall performance regression observed for Spark 4.1.2 .

| Item | Value |
| --- | --- |
| Dataset | `/bigdata/tpcds_data/parquet_100f` |
| Format | Parquet |
| RAPIDS jar | `rapids-4-spark_2.13-26.08.0-SNAPSHOT-cuda13-412-perf.jar` |
| Spark 4.1.1 | `/bigdata/work/tools/spark-4.1.1-bin-hadoop3` |
| Spark 4.1.2 | `/bigdata/work/tools/spark-4.1.2-bin-hadoop3` |
| Runs | 3 per shim |
| Compared runs | Warm runs only: run 2 and run 3 |
| Query status | 618 query JSON records checked, 0 non-Completed |
| Result | No overall performance regression observed for Spark 4.1.2 |

| Metric | Spark 4.1.1 | Spark 4.1.2 | 4.1.2 vs 4.1.1 |
| --- | ---: | ---: | ---: |
| Power run 2 | 462.000s | 467.000s | +1.08% |
| Power run 3 | 489.000s | 472.000s | -3.48% |
| Power run avg | 475.500s | 469.500s | -1.26% |
| Sum of per-query avg times | 473.987s | 468.621s | -1.13% |

### Checklists

Documentation
- [x] Updated for new or modified user-facing features or behaviors
- [ ] No user-facing change

Testing
- [x] Added or modified tests to cover new code paths
- [ ] Covered by existing tests
      (Please provide the names of the existing tests in the PR description.)
- [ ] Not required

Performance
- [x] Tests ran and results are added in the PR description
- [ ] Issue filed with a link in the PR description
- [ ] Not required

### CODE REVIEW

**Comment by firestarman** [comment]
Also depends on the latest private jar being ready on maven repo.

**Comment by firestarman** [comment]
I will launch the premerge once the private jar is ready on the maven repo.

**Comment by firestarman** [comment]
build

**Review by res-life — DISMISSED** [review]
LGTM

**Comment by res-life** [comment]
Sorry, more nits, may be a follow-up:
Add UTs to test boundary values:
add a HyperbolicMathExpressionsSuite.scala under sql-plugin/src/test/spark412/ that calls HyperbolicMathExpressions.acosh/asinh against ColumnVectors with boundary values (LARGE_ACOSH, LARGE_ACOSH - 1e-300, 0.5→NaN, ±LARGE_ASINH, 0.0).
add a MathExpressionsSuite (or extend SparkShimsSuite) in the spark412 test directory that calls GpuAcosh.doColumnar / GpuAsinh.doColumnar against vectors with the exact boundary values: LARGE_ACOSH (≈1.34e154), LARGE_ACOSH - 1, 1.0 (acosh domain boundary), 0.999 (invalid domain → NaN), and Double.PositiveInfinity.

**Comment by firestarman** [comment]
build

**Comment by firestarman** [comment]
build

**Comment by firestarman** [comment]
build

**Comment by firestarman** [comment]
build

File: sql-plugin/src/main/spark330/scala/com/nvidia/spark/rapids/shims/DistributionUtil.scala [hunk 1]
```diff
@@ -38,6 +38,7 @@
 {"spark": "401"}
 {"spark": "402"}
 {"spark": "411"}
+{"spark": "412"}
```
Comment (ttnghia) [inline sql-plugin/src/main/spark330/scala/com/nvidia/spark/rapids/shims/DistributionUtil.scala:41]: Should we investigate a better way to add shim instead of modifying hundreds of files like this each time?
CC @sameerz.

Comment (firestarman) [reply sql-plugin/src/main/spark330/scala/com/nvidia/spark/rapids/shims/DistributionUtil.scala:41]: That's a good suggestion, however it is likely to be another topic.

**Comment by firestarman** [comment]
build

**Comment by firestarman** [comment]
CI is blocked by https://github.com/NVIDIA/cudf-spark/issues/15073

**Comment by firestarman** [comment]
build

**Review by res-life — DISMISSED** [review]
LGTM

**Comment by firestarman** [comment]
build

**Comment by firestarman** [comment]
build

**Comment by firestarman** [comment]
build

File: sql-plugin/src/main/spark412/scala/com/nvidia/spark/rapids/shims/mathExpressions.scala [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+          }
+          withResource(isInvalid) { isInvalid =>
+            withResource(Scalar.fromDouble(Double.NaN)) { nan =>
+              isInvalid.ifElse(nan, largeOrBasic)
+            }
+          }
+        }
+      }
+    }
+  }
```
[the remark this answers is not in the document]
Comment (firestarman) [reply sql-plugin/src/main/spark412/scala/com/nvidia/spark/rapids/shims/mathExpressions.scala]: will do it in a follow-up PR

**Comment by firestarman** [comment]
premerge is blocked by https://github.com/NVIDIA/cudf-spark/issues/15136

**Comment by firestarman** [comment]
build

**Review by res-life — APPROVED** [review]
LGTM

**Comment by firestarman** [comment]
build

**Comment by firestarman** [comment]
build

**Review by NvTimLiu — APPROVED** [review]
LGTM from DevOps point of view for the pom.xml files change
